### PR TITLE
adding possibility to specify resource group for referred virtual net…

### LINKED ISSUE #36624: Increase Azure network interface provisioning flexibility
##### ISSUE TYPE
Feature Idea

 - Bug Report
 - Feature Idea
 - Documentation Report

##### COMPONENT NAME
azure_rm_networkinterface

##### ANSIBLE VERSION

```
N/A
```

##### SUMMARY
The current implementation of the azure_rm_networkinterface does not allow to create a NIC in a resource group R and connect it to a virtual network in a different resource group X. The implementation is assuming that both are in the same resource group and lacks configuration of a separate resource group where the virtual network lives in.

In contrast to that this is already possible with the azure_rm_virtualmachine module which offers a dedicated option `virtual_network_resource_group` for that purpose

##### STEPS TO REPRODUCE

```yaml
- name: Create a network interface that connects to a vnet in a different resource group
  azure_rm_networkinterface:
        name: nic001
        resource_group: Testing
        virtual_network_name: vnet001
        virtual_network_resource_group: TestingNet
        subnet_name: subnet001
```

##### EXPECTED RESULTS

Once implemented, it should be possible to connect NICs to virtual networks in different resource groups, as it's currently already possible with the azure_rm_virtualmachine module

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by zikalino**
…work

##### SUMMARY
Fixes #36624 

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
azure_rm_networkinterface

##### ANSIBLE VERSION
2.4

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py [hunk 1]
```diff
@@ -425,7 +432,10 @@ def exec_module(self, **kwargs):
         # parse the virtual network resource group and name
         virtual_network_dict = parse_resource_id(self.virtual_network_name)
         virtual_network_name = virtual_network_dict.get('name')
-        virtual_network_resource_group = virtual_network_dict.get('resource_group', self.resource_group)
```
Comment (yuwzho) [inline lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py:425]: The origin logic is 
```
dict = parse_virtual_network_id()
virtual_network_resource_group = dict[resource_group] || self.resource_group
```
And now it is changed to 
```
dict = parse_virtual_network_id()
virtual_network_resource_group = dict[resource_group] || self.virtual_network_resource_group || self.resource_group
```

I wonder why this extra logic is needed? Parsing id shares the same behavior like Azure-CLI

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py:425]: dict[resource_group] will work if user specifies entire path/id of virtual_network, but if they want to specify virtual_network from another resource group, specifying virtual_network_resource_group will be simplified way to do this.
This pattern is also implemented in azure_rm_virtualmachine.

Comment (yuwzho) [reply lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py:425]: It is okay, here I would like to make the precedence more clearly.

**Comment by zikalino** [comment]
ready_for_review

File: lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py (hunk 1 above)
Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py:425]: I agree, the documentation should be clearer around the precedence, i.e. `virtual_network_resource_group` overrides `resource_group` and so on.

**Review by jborean93 — COMMENTED** [review]
Just a minor nitpick around documentation and making the precedence clearer. I would also like to see a test to cover this feature to ensure no regressions occur in the future.

**Comment by zikalino** [comment]
@jborean93  added test & updated documentation

**Comment by kyliel** [comment]
@jborean93 , could  you please review this change and see whether it is ok to move forward for PR merge? It also passed the test of the submitter @ingokofler. Thank you.

**Comment by jborean93** [comment]
@zikalino sorry it seems like it has some branch conflicts, are you able to fix them up and I'll get onto the review straight away.

**Comment by zikalino** [comment]
@jborean93 -- resolved

**Comment by zikalino** [comment]
oh no... GitHub created merge commit :-(

**Comment by zikalino** [comment]
ok, now should be rebased

**Review by jborean93 — APPROVED** [review]

**Comment by jborean93** [comment]
rebuild_merge

**Comment by zikalino** [comment]
@jborean93 seems like this one is not merged automatically either....

**Pull request merged**